### DISCUSSION TITLE
What is the difference between recipes, bundles and packages?

### DISCUSSION THREAD

**Discussion by nbro** [symfony/symfony#54895] [Q&A]
If I understand correctly, _bundle_ is a synonym for _package_, but a recipe is something different. It seems that a _recipe_ is a _Symfony Flex_ concept. My understanding is that a recipe is a way to install a bundle (aka package) **and** automatically set certain configuration options for that bundle/package. 

If this is all correct, I don't understand why aren't all packages recipes and why is Symfony Flex not integrated into `composer`. Why isn't the installation of recipes the default behavior? Are there cases where we need just to install bundles without sensible defaults? If yes, can you provide a few examples?

Also, is there a command, which I can type on the terminal, that lists only the recipes, and another that lists only the packages, or a single command that can show if something is a recipe or just a bundle? I couldn't find that in `composer show --help`

**Comment by Trismegiste** [comment]
Short summary :
* Package : a PHP library managed by composer that (generally) goes into `/vendor`. No need for symfony.
* Bundle : a package integrated for Symfony that adds controllers, assets, config files, twig templates...Files goes into `/vendor` too, therefore unversioned in git. A bundle adds a class into the list in `/config/bundles.php` (this is a "default recipe" in a way)
* Recipe : add/edit files outside vendor folder : it could add a config files, edit a docker-compose.yml for example...

**Reply by nbro** [reply]
Can you provide an example of a package, bundle and a recipe? 

If a bundle is a package written specifically for Symfony, then when would you need to install a non-Symfony package in a Symfony project?

What do you mean by "Recipe : add/edit files outside vendor folder"? **Is a recipe a package or not?** Or is it just something that adds config files and it's not a package itself? I **had** already understood that recipes involve config files. My question was more - if recipes are bundles or packages that also set up certain sensible config options, why shouldn't all packages and bundles be recipes?

**Comment by mbabker** [comment]
> If this is all correct, I don't understand why aren't all packages recipes and why is Symfony Flex not integrated into `composer`. Why isn't the installation of recipes the default behavior? Are there cases where we need just to install bundles without sensible defaults? If yes, can you provide a few examples?

Composer is a tool for all of PHP.  Flex is a Composer plugin that hooks into Composer to do some stuff for Symfony applications.  So it can't just be integrated into Composer because many of the things Flex does don't apply to Laravel, Laminas, Yii, or framework-less environments.

> If I understand correctly, _bundle_ is a synonym for _package_, but a recipe is something different. It seems that a _recipe_ is a _Symfony Flex_ concept. My understanding is that a recipe is a way to install a bundle (aka package) **and** automatically set certain configuration options for that bundle/package.

For Composer, there is no difference between a package and a bundle.  A bundle is a Symfony framework concept, the common use cases include integrating a standalone PHP package into a Symfony application or extending Symfony framework features.

A recipe is indeed a feature of the Flex plugin.  It's entirely opt-in, but part of what it offers is the ability to add in opinionated default configurations and files to your Symfony application.

> If a bundle is a package written specifically for Symfony, then when would you need to install a non-Symfony package in a Symfony project?

Because most of the time bundles are integrating PHP packages that don't have Symfony-specific optimizations built right into them.  Two examples of different types of bundles and their relationships with more generic PHP packages:

- The `DoctrineBundle` integrates Doctrine's DBAL and ORM into Symfony applications and adds Symfony-specific functionality for those libraries (i.e. the debug tooling that makes its info available in the web profiler)
- The `LexikJWTAuthenticationBundle` extends Symfony's security system and adds support for JWT-based authentication, and under the hood, it uses several other PHP packages to offer that functionality

**Reply by nbro** [reply]
Thanks for the answer. 

But I think this still doesn't answer the question - In Symfony, when would you need a package which is not a bundle (if ever)? Can you provide an example? I suppose anything that doesn't strictly need to follow the Symfony conventions but that can still be used. Maybe some utility. However, to have a clearer picture, I'd like to know if my assumptions are correct and I'd like to know a specific example of a package, which is not a bundle, that is often used

**Reply by mbabker** [reply]
Neither the Doctrine DBAL nor ORM libraries are Symfony bundles.  Both are very heavily used in the Symfony ecosystem.  There is a dedicated package, the DoctrineBundle, which takes those packages, and integrates them into a Symfony application.

Monolog is a logging library heavily used in the PHP ecosystem.  It is not a Symfony bundle.  There is a MonologBundle that you install to integrate that library into your Symfony application, and it will install Monolog as part of its dependency chain.

You might not directly install Monolog yourself, but if you have the MonologBundle installed, you have the standalone Monolog library installed as well.

**Reply by nbro** [reply]
Thanks again. 

My other question is about the concept of a **recipe**. It's still unclear to me if a recipe is a bundle with the ability of auto-configuring itself when you install it, or if a recipe is something separate from a bundle, which can or not be added to the bundle. According to your explanation, recipes are optional, so I assume they are not packages or bundles, but other code (but what exactly?) you may add to bundles, which allows the bundles to auto-configure themselves. Anyway, it's not clear to me if I should think of recipes as bundles/packages or something else. 

According to the answer [here](https://github.com/symfony/symfony/discussions/54891), the Flex plugin saves the recipes in the `symfony.lock`, but they look like packages, because they have versions, etc. If I look at the file, some of the entries have a `recipe` attribute, not all, which is confusing, because, if `symfony.lock` is meant to contain just recipes, then why don't all entries there have a `recipe` attribute?

**Reply by mbabker** [reply]
A recipe is its own thing.  It's not a bundle, or a package, or an add-on, or anything like that.  In most cases it's just some default configuration to help get you going, maybe some new default files, with the intent of saving you a few moments of copy/pasting some boilerplate code into your application (be it a new file in the `config/packages` path or additions to your `.env` file, as an example).  You can look at the https://github.com/symfony/recipes and https://github.com/symfony/recipes-contrib repos to see where those come from.  Flex has recipes for both Symfony bundles and more generalized PHP libraries (i.e. if you use PHP-CS-Fixer in your application, Flex can automatically create a new default config file for that tool and update your `.gitignore` file based on the recipe info).

The Flex plugin keeps track of what version a package was introduced into an application at, and what version of a recipe was last applied, and stores that in a separate lock file because it is metadata specific to the Flex plugin and not Composer metadata.  The fact that it includes references to packages without recipes really isn't a big deal.

**Reply by nbro** [reply]
So, are you saying that all packages handled by Flex are placed in `symfony.lock`, **even if they don't have recipes** or we don't install recipes for them?

**Reply by kira0269** [reply]
This an example of my symfony.lock:
```json
{
    "artgris/filemanager-bundle": {
        "version": "2.7.1"
    },
    "doctrine/annotations": {
        "version": "2.0",
        "recipe": {
            "repo": "github.com/symfony/recipes",
            "branch": "main",
            "version": "1.10",
            "ref": "64d8583af5ea57b7afa4aba4b159907f3a148b05"
        }
    },
    "friendsofphp/php-cs-fixer": {
        "version": "3.16",
        "recipe": {
            "repo": "github.com/symfony/recipes",
            "branch": "main",
            "version": "3.0",
            "ref": "be2103eb4a20942e28a6dd87736669b757132435"
        },
        "files": [
            ".php-cs-fixer.dist.php"
        ]
    },
    "symfony/debug-bundle": {
        "version": "6.2",
        "recipe": {
            "repo": "github.com/symfony/recipes",
            "branch": "main",
            "version": "5.3",
            "ref": "5aa8aa48234c8eb6dbdd7b3cd5d791485d2cec4b"
        },
        "files": [
            "config/packages/debug.yaml"
        ]
    },
    ...
}
```

As you can see, you will find either:
- bundles with no recipe (artgris/filemanager-bundle)
- packages (not bundles) without recipe (phpcsfixer)
- packages with recipes (doctrine/annotations)
- bundles with recipes (symfony/debug-bundle)

**Reply by nbro** [reply]
It seems that packages can be both in `composer.lock` and `symfony.lock`. For example, in my case, `api-platform/core` appears in both. So, I suppose that composer also keeps tracks of the Flex packages/bundles and Flex, for some reason, just maintains the separate file `symfony.lock` with the packages it maintains too. Maybe the reason is that Flex needs a new syntax to also keep track of other info, including the recipes.

**Reply by mbabker** [reply]
> So, I suppose that composer also keeps tracks of the Flex packages/bundles

No, that's not how it works.

Composer is the package manager, it keeps track of all the dependencies for your application and the versions that those dependencies are running at.  Without Composer, none of it works.

Flex is a Composer plugin adding Symfony-specific functionality to the Composer workflow.  Part of that functionality is in applying recipes.  The metadata the Flex plugin uses for keeping track of those recipes is completely separate from the Composer metadata for keeping track of all of your application's dependencies so that information is tracked in a separate lock file.

The `composer.lock` file and `symfony.lock` are not functionally equivalent and you shouldn't try to tie them together.

**Reply by nbro** [reply]
That's exactly what I wrote in my last comment. Also, you say "No, that's not how it works.", but then you say that composer keeps tracks of all the dependencies, which means it also keeps track of the "flex dependencies".

**Reply by mbabker** [reply]
There isn't a notion of "Flex dependencies" though.  Flex has metadata to keep track of what packages are installed and what recipes it applied, and that is the metadata it writes to `symfony.lock`.  But in no way is Flex involved with managing the application's dependencies; that is all Composer.  You're trying to tie the two systems together when in reality they're only loosely related because Flex (because of the features it provides) has to run as a Composer plugin.  Truthfully, all of the recipe stuff could be extracted and run in its own standalone CLI tool, but then all the DX convenience the plugin provides is basically wiped out.

**Reply by nbro** [reply]
When I said flex dependencies, I didn't mean that flex does the dependency resolution. In fact, I used it in quotes. But flex surely does something with dependencies or packages or whatever you want to call them. My questions would not even exist if this was not the case.
